### PR TITLE
Prohibit DML in function bodies (for now)

### PR DESCRIPTION
**Pull request by elprans**
Data modification in functions has a number of problems:

- Effects of DML inside a function is invisible to the calling query,
  which is confusing.
- The expression evaluation semantics assumes that functions have no
  side effects and so the number of invocations (0 or >1) is not
  defined.

All these problems are present in PostgreSQL as well.

To avoid surprising behavior or potential future compatibility breaks, 
prohibit DML in functions altogether until a satisfactory solution to 
mutating functions is implemented.

To preserve compatibility with existing databases that have mutating 
functions, introduce the `allow_dml_in_functions` configuration parameter, 
which enables historical behavior of allowing DML in functions and is set
automatically when a dump file from a pre-beta instance is restored.

### CODE REVIEW

File: edb/lib/cfg.edgeql [hunk 1]
```diff
@@ -23,6 +23,7 @@ CREATE ABSTRACT INHERITABLE ANNOTATION cfg::backend_setting;
 CREATE ABSTRACT INHERITABLE ANNOTATION cfg::internal;
 CREATE ABSTRACT INHERITABLE ANNOTATION cfg::requires_restart;
 CREATE ABSTRACT INHERITABLE ANNOTATION cfg::system;
+CREATE ABSTRACT INHERITABLE ANNOTATION cfg::compilation;
```
Comment (1st1) [inline edb/lib/cfg.edgeql:26]: I don't quite like the name, it's gonna be confusing in introspection :/ Maybe `cfg::required_for_compiler`?

File: edb/server/compiler/compiler.py [hunk 2]
```diff
@@ -107,9 +110,9 @@ class CompileContext:
     bootstrap_mode: bool = False
     internal_schema_mode: bool = False
     standalone_mode: bool = False
+    system_config: immutables.Map = EMPTY_MAP
```
Comment (1st1) [inline edb/server/compiler/compiler.py:114]: Why is this here and not in CompilerConnectionState?

File: edb/server/compiler/compiler.py [hunk 3]
```diff
@@ -1898,18 +1914,26 @@ def compile(
...
         state = dbstate.CompilerConnectionState(
-            user_schema,
-            global_schema,
-            sess_modaliases,
-            sess_config,
-            reflection_cache,
+            user_schema=user_schema,
+            global_schema=global_schema,
+            modaliases=sess_modaliases,
+            session_config=sess_config,
```
Comment (1st1) [inline edb/server/compiler/compiler.py:1941]: I see you've also added it to CompilerConnectionState, not sure it's needed in CompileContext

File: tests/schemas/dump02_default.esdl [hunk 4]
```diff
@@ -69,3 +69,10 @@ type Tree {
     link children := .<parent[IS Tree];
     property child_vals := .children.val;
 }
+
```
Comment (1st1) [inline tests/schemas/dump02_default.esdl:72]: Is adding tests/dumps/dump02/1_0a8.dump intentional? We don't have a8, so it's confusing.

**Review by 1st1 — APPROVED** [review]
LGTM. I looked at changes in server/compiler, server/compiler_pool, server/dbview, etc. rather thoroughly. There are a couple of nit comments though.

File: tests/schemas/dump02_default.esdl (hunk 4 above)
Comment (elprans) [reply tests/schemas/dump02_default.esdl:72]: I can't generate it from a7 due to migration changes, so a8 is the best approximation of state between a7 and b1

File: edb/server/compiler/compiler.py (hunk 2 above)
Comment (elprans) [reply edb/server/compiler/compiler.py:114]: System config isn't transactional, so it seemed unnecessary to do all the pointless transaction tracking for it in `CompilerConnectionState`.  That said, it might still worth it for consistency, I guess.

File: edb/lib/cfg.edgeql (hunk 1 above)
Comment (elprans) [reply edb/lib/cfg.edgeql:26]: How about `affects_compilation`?

**Comment by fantix** [comment]
I met an issue below with RDS and bin-searching located [this commit](3881723c4ebe1533d40a89a11b5e7e81fc243faa) in this PR:

```
$ edgedb-server --postgres-dsn=postgres://...@...us-east-2.rds.amazonaws.com:5432/postgres
INFO 43594 2021-02-25 20:57:19,108 edb.server: EdgeDB server (1.0-alpha.8+dev.5438.g3881723c4) starting in DEV mode.
================================================================================ Exception occurred: structure of query does not match function result type ================================================================================

1. edb.server.pgcon.errors.BackendError: structure of query does not match function result type

---------------------------------------------------------------------------------------------------------------- Traceback -----------------------------------------------------------------------------------------------------------------

    bin/edgedb-server, line 33, in <module>
        > sys.exit(load_entry_point('edgedb-server', 'console_scripts', 'edgedb-server')())
    edb/server/main.py, line 711, in main_dev
        > main()
    lib/python3.8/site-packages/click/core.py, line 829, in __call__
        > return self.main(*args, **kwargs)
    lib/python3.8/site-packages/click/core.py, line 782, in main
        > rv = self.invoke(ctx)
    lib/python3.8/site-packages/click/core.py, line 1066, in invoke
        > return ctx.invoke(self.callback, **ctx.params)
    lib/python3.8/site-packages/click/core.py, line 610, in invoke
        > return callback(*args, **kwargs)
    edb/server/main.py, line 706, in main
        > server_main(**kwargs)
    edb/server/main.py, line 695, in server_main
        > run_server(ServerConfig(**kwargs))
    edb/server/main.py, line 345, in run_server
        > _run_server(cluster, args, runstate_dir, internal_runstate_dir)
    edb/server/main.py, line 232, in _run_server
        > loop.run_until_complete(ss.init())
    uvloop/loop.pyx, line 1494, in uvloop.loop.Loop.run_until_complete

    edb/server/server.py, line 221, in init
        > sys_config = await self.load_sys_config()
    edb/server/server.py, line 306, in load_sys_config
        > sys_config_json = await syscon.parse_execute_json(
    edb/server/pgcon/pgcon.pyx, line 497, in parse_execute_json
        > return await self._parse_execute_json(
    edb/server/pgcon/pgcon.pyx, line 448, in _parse_execute_json
        > await self._parse_execute_to_buf(
    edb/server/pgcon/pgcon.pyx, line 431, in _parse_execute_to_buf
        > raise error

edb.server.pgcon.errors.BackendError: structure of query does not match function result type
```

I'm still trying to look into why this happened, but any help would be appreciated.

**Comment by elprans** [comment]
I know what the issue is, will fix momentarily.

**Comment by elprans** [comment]
Fix: #2275

**Comment by fantix** [comment]
Oh wow that's fast, thanks! I can confirm that this is fixed.
